### PR TITLE
Tony/bur-33-task-11-develop-unit-and-integration-test-suite

### PR DESCRIPTION
**Pull request by tonyblank**
- Introduced additional markers in `pytest.ini` for authentication, resilience, and performance tests to enhance test organization and execution.
- This update aims to facilitate better categorization of tests and improve overall test suite maintainability.

## Summary by CodeRabbit

- **New Features**
  - Introduced comprehensive integration tests for real GitHub API covering authentication flows, end-to-end workflows, network resilience, rate limiting, error recovery, and performance monitoring.
  - Added fixtures and utilities to support integration testing with real APIs, including network simulation, performance monitoring, cleanup helpers, and conditional slow test skipping.
  - Enhanced test data generation and performance baseline fixtures for dynamic and performance-related tests.
  - Added extensive unit tests for utility functions, model edge cases, validators, formatters, CLI commands (fetch, reply, resolve, schema), and services covering input validation, error handling, output formatting, and integration.
  - Applied consistent pytest markers across unit and integration test classes for improved test categorization and selective execution.
  - Added output format support ("pretty" and "json") to pull request selection and fetch services, improving CLI interaction modes and output routing.
  - Added detailed unit tests for CLI core, context handling, error management, environment variable parsing, and main entry point behavior.
  - Introduced new tests for format selection utilities and schema validator edge cases.

- **Bug Fixes**
  - Improved test output readability by setting a default traceback style for test runs.
  - Fixed error message formatting for clearer and consistent exception string conversion.

- **Documentation**
  - Added detailed documentation for the test framework and real API integration tests, including setup instructions, test categorization, and best practices.
  - Added READMEs for unit command tests and real API integration tests.

- **Chores**
  - Updated pytest configuration for marker definitions, test collection, and path settings.
  - Enhanced CI workflow environment with GitHub token for authenticated test runs.
  - Revised example environment variables to focus on testing and API interaction parameters.
  - Added `python-dotenv` dependency for environment variable management.
  - Simplified development dependencies in requirements and pyproject.toml by removing many pytest plugins and adding system utilities.
  - Updated CI workflows to separate integration tests and adjust caching and install commands.
  - Added new pre-commit checks for trailing whitespace and end-of-file fixes in CI scripts.
  - Removed deprecated files and updated Makefile to unify commands under `uv` tool.

- **Refactor**
  - Renamed validation function related to reply content warnings for clarity.
  - Adjusted import paths for node ID validation functions in service code.
  - Routed user-facing output to standard error in JSON mode to avoid polluting output streams.
  - Modified fetch service and PR selector to accept and handle output format parameter for better mode distinction.
  - Modernized type annotations across codebase to use Python 3.9+ built-in generic types (`list`, `dict`, `tuple`) instead of `typing` imports.
  - Simplified control flow by removing redundant else blocks and pass statements.
  - Standardized exception message formatting to use `{e!s}` for consistent string conversion.

### CODE REVIEW

File: tests/integration/conftest.py [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+
+    except subprocess.TimeoutExpired:
+        pytest.skip("GitHub CLI commands timed out")
+    except FileNotFoundError:
+        pytest.skip("GitHub CLI (gh) not found in PATH")
+    except json.JSONDecodeError:
+        pytest.skip("Invalid JSON response from GitHub API")
+    except Exception as e:
+        pytest.skip(f"GitHub API health check failed: {e}")
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/integration/conftest.py:103]: Fixed the UnboundLocalError by moving the `json` import to the top of the `github_api_health_check` function before the try block. This ensures `json` is always available when `json.JSONDecodeError` is caught in the exception handler, eliminating the potential UnboundLocalError that would occur if an exception happened before line 85 where the import was previously located.

File: tests/integration/cli/test_fetch_cli.py [hunk 2]
```diff
@@ -3,11 +3,14 @@
...
 from unittest.mock import Mock, patch
 
+import pytest
 from click.testing import CliRunner
 
 from toady.cli import cli
 
 
+@pytest.mark.integration
+@pytest.mark.cli
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/integration/cli/test_fetch_cli.py:13]: Fixed the pytest marker warnings by correcting the pytest.ini configuration format from `[tool:pytest]` to `[pytest]`. The markers 'integration' and 'cli' were already declared in the pytest.ini file but weren't being recognized due to the incorrect section header format. This resolves the UnknownMarkWarning messages in CI.

File: pyproject.toml [hunk 3]
```diff
@@ -42,34 +42,22 @@ dev = [
...
     "pre-commit>=3.4.0",
-    "tox>=4.0.0",
     "build>=1.0.0",
     "twine>=4.0.0",
     "tomli>=2.0.0;python_version<'3.11'",
-    "coverage[toml]>=7.3.0",
+    "python-dotenv>=1.1.0",
+    "psutil>=5.9.0",
+    "types-click>=7.1.0",
 ]
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply pyproject.toml:55]: Fixed the types-click version mismatch by updating from 7.1.0 to 8.1.0 in pyproject.toml:54. This resolves the compatibility issue with Click ≥8.1 and prevents mypy from throwing type-checking errors. The new version spec 'types-click>=8.1.0' ensures we get stubs that match our Click version, maintaining proper type safety across the codebase.

File: tests/unit/commands/test_schema.py [hunk 4]
```diff
@@ -0,0 +1,1230 @@
...
+        """Test validation when SchemaValidationError is raised during init."""
+        mock_validator_class.side_effect = SchemaValidationError(
+            "Schema invalid", suggestions=["Fix syntax"]
+        )
+
+        result = runner.invoke(cli, ["schema", "validate"])
+
+        assert result.exit_code == 1
+        assert "Configuration error initializing schema validator" in result.output
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/commands/test_schema.py:375]: Fixed the test assertion mismatch by updating the expected error message format in test_schema.py. The tests were expecting 'Configuration error initializing schema validator' but the actual CLI output includes the 'Error: ' prefix from ToadyError handling. Updated all 4 assertion locations (lines 221, 374, 530, 704) to expect 'Error: Configuration error initializing schema validator' to match the actual command output format.

File: tests/unit/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+    @patch("toady.cli.cli")
+    @patch("toady.cli.handle_error")
+    def test_main_debug_environment_access(
+        self, mock_handle_error, mock_cli, mock_env_get
+    ):
+        """Test that main() correctly accesses debug environment variable."""
+        test_error = ToadyError("Test error")
+        mock_cli.side_effect = test_error
+        mock_env_get.return_value = "1"
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/test_cli.py:433]: Fixed the invalid patch target by replacing @patch('os.environ.get') with @patch.dict('os.environ', {'TOADY_DEBUG': '1'}) in test_cli.py:423. The original approach tried to patch a non-existent module 'os.environ' causing test collection to fail. Using patch.dict properly mocks the environment variable directly, which is more appropriate for testing environment variable access. This approach follows the pattern used in other tests and avoids import errors.

File: tests/unit/commands/test_schema.py [hunk 6]
```diff
@@ -0,0 +1,1238 @@
...
+        """Test validation when schema fetch fails with connection error."""
+        mock_validator = Mock()
+        mock_validator.fetch_schema.side_effect = ConnectionError("Network timeout")
+        mock_validator_class.return_value = mock_validator
+
+        result = runner.invoke(cli, ["schema", "validate"])
+
+        assert result.exit_code == 1
+        assert "Network error fetching GitHub schema" in result.output
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/commands/test_schema.py:236]: Fixed the prefix mismatch. Updated all assertions to expect 'Error: Network error fetching GitHub schema' instead of the bare string to match the CLI error handler behavior that prepends 'Error:' to user-visible messages.

File: tests/unit/commands/test_schema.py [hunk 7]
```diff
@@ -0,0 +1,1238 @@
...
+        """Test validation when schema fetch fails with file operation error."""
+        mock_validator = Mock()
+        mock_validator.fetch_schema.side_effect = OSError("Disk full")
+        mock_validator_class.return_value = mock_validator
+
+        result = runner.invoke(cli, ["schema", "validate"])
+
+        assert result.exit_code == 1
+        assert "File operation error during schema fetch" in result.output
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/commands/test_schema.py:262]: Fixed the missing Error: prefix. Updated both assertions to expect 'Error: File operation error during schema fetch' instead of the bare string to maintain consistency with other error message assertions in the test suite.

File: src/toady/formatters/format_interfaces.py [hunk 8]
```diff
@@ -153,21 +147,20 @@ def _safe_serialize(self, obj: Any) -> Any:
...
             return {key: self._safe_serialize(value) for key, value in obj.items()}
-        elif hasattr(obj, "__dict__"):
+        if hasattr(obj, "__dict__"):
             return self._safe_serialize(obj.__dict__)
-        else:
-            # For primitive types and other serializable objects
-            try:
-                import json
+        # For primitive types and other serializable objects
+        try:
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/formatters/format_interfaces.py]: Fixed the to_dict callable check. Added verification that to_dict is callable before invoking it, and wrapped the call in a try-except block to safely fall back to str(obj) if the to_dict method raises an exception.

File: tests/integration/real_api/test_performance_monitoring.py [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+                "fetch",
+                "--pr",
+                str(pr_number),
+                "--resolved",  # Include resolved threads for larger dataset
+                "--limit",
+                "1000",
+                "--format",
+                "json",
+            ],
+        )
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/integration/real_api/test_performance_monitoring.py]: Fixed the --limit 1000 values that exceed the maximum allowed limit of 100. Updated both instances in the performance tests to use --limit 100 instead to comply with the ReviewThreadQueryBuilder and PullRequestQueryBuilder validation constraints.

File: src/toady/parsers/graphql_queries.py [hunk 10]
```diff
@@ -38,7 +38,7 @@ def _validate_cursor(cursor: str) -> bool:
     try:
         base64.b64decode(cursor, validate=True)
     except Exception as e:
-        raise ValueError(f"Invalid Base64 cursor: {str(e)}") from e
+        raise ValueError(f"Invalid Base64 cursor: {e!s}") from e
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply src/toady/parsers/graphql_queries.py]: Fixed the cursor validation docstring vs implementation mismatch. Changed the empty cursor check to raise ValueError('Cursor cannot be empty') instead of returning False, ensuring the behavior matches the docstring that states the function raises ValueError for invalid cursors.

File: .github/workflows/ci-uv.yml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        uv run mypy src
+
+    - name: Run tests
+      run: |
+        uv run pytest --cov=src/toady --cov-report=xml
+
+    - name: Upload coverage
+      uses: codecov/codecov-action@v3
+      with:
+        file: ./coverage.xml
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply .github/workflows/ci-uv.yml:44]: Updated codecov action from v3 to v5. The v3 version is end-of-life and flagged by action-lint. Updated to the latest v5 version to ensure continued support and avoid deprecation warnings.

File: .github/workflows/ci-uv.yml [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        uv run mypy src
+
+    - name: Run tests
+      run: |
+        uv run pytest --cov=src/toady --cov-report=xml
+
+    - name: Upload coverage
+      uses: codecov/codecov-action@v5
+      with:
+        file: ./coverage.xml
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply .github/workflows/ci-uv.yml:44]: Fixed the invalid Codecov input by changing 'file' to 'files' to match the supported parameter name in codecov/codecov-action@v5.

File: tests/integration/real_api/test_end_to_end_workflows.py [hunk 13]
```diff
@@ -0,0 +1,506 @@
...
+
+            # Re-fetch to get the new comment
+            result = integration_cli_runner.invoke(
+                cli, ["fetch", "--pr", str(pr_number), "--format", "json"]
+            )
+
+            assert result.exit_code == 0
+            threads_data = json.loads(result.output)
+
+        if not threads_data:
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/integration/real_api/test_end_to_end_workflows.py:103]: Fixed the inconsistent output handling by using the same raw_out pattern (getattr(result, "stdout", result.output)) for consistent Click 7/8 compatibility.

File: tests/integration/real_api/test_network_resilience.py [hunk 14]
```diff
@@ -0,0 +1,533 @@
...
+        large_response_duration = performance_monitor.stop_timing()
+
+        # Should handle large responses within reasonable time
+        performance_monitor.assert_performance_threshold("large_response_fetch", 60.0)
+
+        if result.exit_code == 0:
+            try:
+                threads = json.loads(result.output)
+                assert isinstance(threads, list)
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/integration/real_api/test_network_resilience.py:118]: Fixed the test logic to assert command success before checking performance thresholds. Now it properly fails if the CLI command fails, preventing silent failures from passing tests.

File: tests/unit/services/test_github_service.py [hunk 15]
```diff
@@ -546,3 +625,198 @@ def test_exception_messages(self) -> None:
...
+        """Test error handling in review ID lookup."""
+        service = GitHubService()
+
+        # Mock the private method to test error paths
+        mock_get_review.return_value = None
+
+        # This should be called through post_reply when review_id can't be determined
+        result = service._get_review_id_for_comment("IC_kwDOABcD12MAAAABcDE3fg")
+        assert result is None
+
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/services/test_github_service.py:769]: Fixed the test by removing the patch of the method under test and instead mocking execute_graphql_query to return a response with no review field, allowing the real implementation to be tested.

File: tests/unit/test_error_handling.py [hunk 16]
```diff
@@ -130,10 +130,17 @@ def test_format_toady_error_with_suggestions(self):
...
+        formatted = ErrorMessageFormatter.format_error(error)
         assert "❌ An unexpected error occurred" in formatted
-        assert "Error details: Unexpected error" in formatted
+        assert "Error details: Unexpected error" not in formatted
+
+        # Test with debug mode enabled
+        with patch.dict("os.environ", {"TOADY_DEBUG": "1"}):
+            formatted_debug = ErrorMessageFormatter.format_error(error)
+            assert "❌ An unexpected error occurred" in formatted_debug
+            assert "Error details: Unexpected error" in formatted_debug
```
[the remark this answers is not in the document]
Comment (tonyblank) [reply tests/unit/test_error_handling.py:144]: Made the test deterministic by wrapping the no-debug branch in patch.dict with clear=True to ensure TOADY_DEBUG is not set, removing dependency on external environment variables.
